`div`: port to structured

bdhirsh opened the pull request:
Porting `div` to structured.

One weird thing to call out with div: It has an overload, `div.Tensor_mode`, which uses different TensorIterator settings depending on this input (the "mode" argument that you pass to it). So I ended up switching on the mode inside of the meta function to determine which TensorIterator builder to use.


Stack from [ghstack](https://github.com/ezyang/ghstack):
* #54085 fix torch.pow type promotion issue
* **#53680 port div to structured**
* #53679 port sub to structured, fix sub.Scalar bug
* #53669 port at::pow to structured, fix pow bug with mixed cpu/cuda tensors
* #52692 port at::mul to structured

Differential Revision: [D27029819](https://our.internmc.facebook.com/intern/diff/D27029819)

bdhirsh committed:
port div to structured

[ghstack-poisoned]

bdhirsh committed:
Update on "port div to structured"




[ghstack-poisoned]

bdhirsh committed:
Update on "port div to structured"




[ghstack-poisoned]

bdhirsh committed:
Update on "port div to structured"




[ghstack-poisoned]

bdhirsh committed:
Update on "port div to structured"

Porting `div` to structured.

One weird thing to call out with div: It has an overload, `div.Tensor_mode`, which uses different TensorIterator settings depending on this input (the "mode" argument that you pass to it). So I ended up switching on the mode inside of the meta function to determine which TensorIterator builder to use.





[ghstack-poisoned]

bdhirsh committed:
Update on "port div to structured"

Porting `div` to structured.

One weird thing to call out with div: It has an overload, `div.Tensor_mode`, which uses different TensorIterator settings depending on this input (the "mode" argument that you pass to it). So I ended up switching on the mode inside of the meta function to determine which TensorIterator builder to use.





[ghstack-poisoned]

ezyang commented on aten/src/ATen/native/BinaryOps.cpp:
```diff
@@ -34,6 +34,24 @@ TORCH_META_FUNC2(mul, Tensor) (
+  build_binary_float_op(maybe_get_output(), self, other);
+}
+
+TORCH_META_FUNC2(div, Tensor_mode) (const Tensor& self, const Tensor& other, std::string rounding_mode) {
```
ew, we really should get this to stop passing in strings LOL

bdhirsh committed:
Update on "port div to structured"

Porting `div` to structured.

One weird thing to call out with div: It has an overload, `div.Tensor_mode`, which uses different TensorIterator settings depending on this input (the "mode" argument that you pass to it). So I ended up switching on the mode inside of the meta function to determine which TensorIterator builder to use.



Differential Revision: [D27029819](https://our.internmc.facebook.com/intern/diff/D27029819)

[ghstack-poisoned]

bdhirsh committed:
Update on "port div to structured"

Porting `div` to structured.

One weird thing to call out with div: It has an overload, `div.Tensor_mode`, which uses different TensorIterator settings depending on this input (the "mode" argument that you pass to it). So I ended up switching on the mode inside of the meta function to determine which TensorIterator builder to use.



Differential Revision: [D27029819](https://our.internmc.facebook.com/intern/diff/D27029819)

[ghstack-poisoned]

bdhirsh committed:
Update on "port div to structured"

Porting `div` to structured.

One weird thing to call out with div: It has an overload, `div.Tensor_mode`, which uses different TensorIterator settings depending on this input (the "mode" argument that you pass to it). So I ended up switching on the mode inside of the meta function to determine which TensorIterator builder to use.



Differential Revision: [D27029819](https://our.internmc.facebook.com/intern/diff/D27029819)

[ghstack-poisoned]

bdhirsh committed:
Update on "port div to structured"

Porting `div` to structured.

One weird thing to call out with div: It has an overload, `div.Tensor_mode`, which uses different TensorIterator settings depending on this input (the "mode" argument that you pass to it). So I ended up switching on the mode inside of the meta function to determine which TensorIterator builder to use.



Differential Revision: [D27029819](https://our.internmc.facebook.com/intern/diff/D27029819)

[ghstack-poisoned]

bdhirsh committed:
Update on "port div to structured"

Porting `div` to structured.

One weird thing to call out with div: It has an overload, `div.Tensor_mode`, which uses different TensorIterator settings depending on this input (the "mode" argument that you pass to it). So I ended up switching on the mode inside of the meta function to determine which TensorIterator builder to use.



Differential Revision: [D27029819](https://our.internmc.facebook.com/intern/diff/D27029819)

[ghstack-poisoned]

bdhirsh committed:
Update on "port div to structured"

Porting `div` to structured.

One weird thing to call out with div: It has an overload, `div.Tensor_mode`, which uses different TensorIterator settings depending on this input (the "mode" argument that you pass to it). So I ended up switching on the mode inside of the meta function to determine which TensorIterator builder to use.



Differential Revision: [D27029819](https://our.internmc.facebook.com/intern/diff/D27029819)

[ghstack-poisoned]

bdhirsh committed:
Update on "port div to structured"

Porting `div` to structured.

One weird thing to call out with div: It has an overload, `div.Tensor_mode`, which uses different TensorIterator settings depending on this input (the "mode" argument that you pass to it). So I ended up switching on the mode inside of the meta function to determine which TensorIterator builder to use.



Differential Revision: [D27029819](https://our.internmc.facebook.com/intern/diff/D27029819)

[ghstack-poisoned]
